### PR TITLE
Add the `AudioClassificationPipeline`

### PR DESCRIPTION
**Pull request by anton-l**
# What does this PR do?

This adds the audio classification pipeline needed for `Wav2Vec2ForSequenceClassification` and others (see #13153).

The implementation is mostly based on `ImageClassificationPipeline` with `ffmpeg` audio file loading borrowed from `AutomaticSpeechRecognitionPipeline` by @Narsil 

Once merged, model cards like https://hf.co/superb/hubert-base-superb-ks should be able to have an `audio-classification` inference widget.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @Narsil

### CODE REVIEW

File: tests/test_pipelines_audio_classification.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                {"score": 0.0073, "label": "up"},
+                {"score": 0.0064, "label": "_unknown_"},
+                {"score": 0.0015, "label": "down"},
+            ],
+        )
+
+        # test with a local file
+        dataset = datasets.load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
+        filename = dataset[0]["file"]
+        output = audio_classifier(filename, top_k=2)
```
Comment (anton-l) [inline tests/test_pipelines_audio_classification.py:74]: How was this passing for `AutomaticSpeechRecognitionPipeline`? The runner doesn't seem to have `ffmpeg` installed to load the file.

File: src/transformers/models/auto/modeling_auto.py [hunk 2]
```diff
@@ -439,6 +439,14 @@
     ]
 )
 
+MODEL_FOR_AUDIO_CLASSIFICATION_MAPPING_NAMES = OrderedDict(
+    [
+        # Model for Audio Classification mapping
+        ("wav2vec2", "Wav2Vec2ForSequenceClassification"),
```
Comment (anton-l) [inline src/transformers/models/auto/modeling_auto.py:445]: Should we quickly rename them to `...ForAudioClassification`? To avoid confusion when using pipelines and be more in line with `...ForImageClassification`

File: tests/test_pipelines_audio_classification.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        )
+
+    @require_torch
+    @require_datasets
+    def test_small_model_pt(self):
+        import datasets
+
+        model = "superb/wav2vec2-base-superb-ks"
+        tokenizer = PreTrainedTokenizer()
+        audio_classifier = pipeline("audio-classification", model=model, tokenizer=tokenizer)
```
Comment (anton-l) [inline tests/test_pipelines_audio_classification.py:55]: Any way to work around having an empty `tokenizer` here? Otherwise the `pipeline()` tries to load `Wav2Vec2CTCTokenizer` by default, since it's defined in the config mapper for `wav2vec2` and `hubert`.

File: src/transformers/pipelines/audio_classification.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+if is_torch_available():
+    import torch
+
+    from ..models.auto.modeling_auto import MODEL_FOR_AUDIO_CLASSIFICATION_MAPPING
+
+logger = logging.get_logger(__name__)
+
+
+def ffmpeg_read(bpayload: bytes, sampling_rate: int) -> np.array:
```
Comment (anton-l) [inline src/transformers/pipelines/audio_classification.py:37]: Kinda want to move it to helpers/utils somewhere, to reuse in both ASR and classification

Comment (patrickvonplaten) [reply src/transformers/pipelines/audio_classification.py:37]: Think it's fine for now - @Narsil what do you think?

File: src/transformers/pipelines/audio_classification.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        processed = self.feature_extractor(
+            inputs, sampling_rate=self.feature_extractor.sampling_rate, return_tensors="pt"
+        )
+        processed = self.ensure_tensor_on_device(**processed)
+
+        with torch.no_grad():
+            outputs = self.model(**processed)
+
+            probs = outputs.logits.softmax(-1)
+            scores, ids = probs.topk(top_k)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/audio_classification.py:150]: why do we take `topk` here if we only care about the best score, label?

File: tests/test_pipelines_audio_classification.py (hunk 3 above)
Comment (patrickvonplaten) [reply tests/test_pipelines_audio_classification.py:55]: yeah for `"audio-classification"` we never need a tokenizer so we should change this here. The following:

```
 model = "superb/wav2vec2-base-superb-ks"
audio_classifier = pipeline("audio-classification", model=model)
```

doesn't work?

Comment (anton-l) [reply tests/test_pipelines_audio_classification.py:55]: Nope, an empty tokenizer falls into [this branch](https://github.com/huggingface/transformers/blob/master/src/transformers/pipelines/__init__.py#L447) which picks up `Wav2Vec2CTCTokenizer` linked in [tokenization_auto](https://github.com/huggingface/transformers/blob/c4ecd234f250cc6272812505b4a9b6cd1e29a816/src/transformers/models/auto/tokenization_auto.py#L171)

Comment (patrickvonplaten) [reply tests/test_pipelines_audio_classification.py:55]: Think it won't be that easy to **not** load a tokenizer for now. We could do an ugly: ``if task_class in ....`` -> don't load tokenizer here: https://github.com/huggingface/transformers/blob/42f359d015aee3835490bdcfa20df657a4d97049/src/transformers/pipelines/__init__.py#L444 but that's not the cleanest solution I think...maybe for now we can just silently load a tokenizer that won't be used. @Narsil what do you think?

File: src/transformers/models/auto/modeling_auto.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/auto/modeling_auto.py:445]: I think it's fine. IMO `ForSequenceClassification` is totally fine...(maybe someone wants to use the model for stock market prediction, classification, ....). We'll do a release tomorrow so not sure we can change it after anymore

File: src/transformers/pipelines/audio_classification.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+        if self.framework != "pt":
+            raise ValueError(f"The {self.__class__} is only available in PyTorch.")
+
+        self.check_model_type(MODEL_FOR_AUDIO_CLASSIFICATION_MAPPING)
+
+    def __call__(
+        self,
+        inputs: Union[np.ndarray, bytes, str],
+        top_k: Optional[int] = 5,
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/audio_classification.py:104]:
```suggestion
        top_k: int = 5,
```

If top_k doesn't default to `None` it's not Optional IMO

File: tests/test_pipelines_audio_classification.py (hunk 1 above)
Comment (patrickvonplaten) [reply tests/test_pipelines_audio_classification.py:74]: The tests are classified as "slow" which means they are only run once a day here: https://github.com/huggingface/transformers/blob/42f359d015aee3835490bdcfa20df657a4d97049/.github/workflows/self-scheduled.yml#L70

Maybe we can do the same here: We have one "fast" test where we just classify a dummy 0 waveform: 

```
waveform = np.zeros((34000,))
```

which doesn't need `ffmpeg` and the other tests (those that you have right now) will be marked as slow

File: tests/test_pipelines_audio_classification.py (hunk 3 above)
Comment (patrickvonplaten) [reply tests/test_pipelines_audio_classification.py:55]: @anton-l - this should still be fine no? IMO it's not that problematic for now if a "useless" tokenizer is initialized (it's not pretty, but changing the general behavior here might require some more changes)

Comment (anton-l) [reply tests/test_pipelines_audio_classification.py:55]: Yeah, should be fine for demo purposes for now

File: src/transformers/pipelines/audio_classification.py (hunk 6 above)
Comment (anton-l) [reply src/transformers/pipelines/audio_classification.py]: Good point, made it `None` by default.

File: tests/test_pipelines_audio_classification.py (hunk 1 above)
Comment (anton-l) [reply tests/test_pipelines_audio_classification.py:79]: Marked this test case as `@slow` and switched to a dummy waveform for the blanket test case above (`run_pipeline_test()`). Both are overrides of `PipelineTestCaseMeta`, but work as expected now.

File: src/transformers/pipelines/audio_classification.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/pipelines/audio_classification.py:37]: I think we could do with `pipelines/{text/audio/vision}` subdirectories, with each getting their pipelines and utils.
`load_image`/`load_audio` also belong in there.

It's better leaving that for a follow up  PR I think though (let's keep this one nice and small)

File: tests/test_pipelines_audio_classification.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        audio = np.zeros((34000,))
+        output = audio_classifier(audio)
+        # by default a model is initialized with num_labels=2
+        self.assertEqual(
+            output,
+            [
+                {"score": ANY(float), "label": ANY(str)},
+                {"score": ANY(float), "label": ANY(str)},
+            ],
+        )
```
Comment (Narsil) [inline tests/test_pipelines_audio_classification.py:50]: I would add a simple test to check that top_k is honored (so top_k = 1, if the configs only defined 2 labels)

File: tests/test_pipelines_audio_classification.py (hunk 1 above)
Comment (Narsil) [reply tests/test_pipelines_audio_classification.py:79]: If you are testing that local files loading is working, it does not belong in `small_model_pt`, that test is reserved for checking values (and checking they don't drift as we make changes to pipelines).

For checking functionality `top_k` / loading from file, we need to run them in `run_pipeline_test` as it's called with every possible model/tokenizer/feature_extractor supported. Some models will trigger errors that others don't.

**Review by Narsil — APPROVED** [review]
LGTM overall

The biggest issue I have is with being forced to use a dummy tokenizer to be able to load with `pipeline(...)`.

`pipeline` function is all about being magic and working in almost all conditions so I definitely think we should enable it. However I am very fine to leave that as a follow up (but it we do we need a comment within the test saying that we plan to enable a simpler call, and that it needs to be changed later).

Thanks for this @anton-l !

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me now!

**Review by Narsil — CHANGES_REQUESTED** [review]
Sorry, just realized I made a mistake.

`test_small_model_pt` MUST be a fast test, but we hsould use a dummy (small) model.

It's necessary to check the values don't drift.

The test you wrote seems like `test_large_model_pt` which is also valuable, but is slow.

**Review by Narsil — APPROVED** [review]
LGTM !

File: src/transformers/pipelines/audio_classification.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        """
+        if isinstance(inputs, str):
+            with open(inputs, "rb") as f:
+                inputs = f.read()
+
+        if isinstance(inputs, bytes):
+            inputs = ffmpeg_read(inputs, self.feature_extractor.sampling_rate)
+
+        assert isinstance(inputs, np.ndarray), "We expect a numpy ndarray as input"
+        assert len(inputs.shape) == 1, "We expect a single channel audio input for AudioClassificationPipeline"
```
Comment (LysandreJik) [inline src/transformers/pipelines/audio_classification.py:137]: We're trying to move away from `assert` statements, would be great to replace these with `ValueError`s

File: tests/test_pipelines_audio_classification.py (hunk 3 above)
Comment (LysandreJik) [reply tests/test_pipelines_audio_classification.py:107]: Agree with @Narsil

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks for working on this @anton-l
